sw/device/silicon_creator/lib/boot_data.c

```diff
@@ -560,3 +507,50 @@ rom_error_t boot_data_write(const boot_data_t *boot_data) {
 
   return kErrorOk;
 }
+
+/**
+ * Shares for producing the `error` value in `boot_data_check()`. First 8
+ * shares are generated using the `sparse-fsm-encode` script while the last
+ * share is `kErrorOk ^ kBootDataInvalid ^ kBootDataIdentifier ^
+ * kCheckShares[0] ^ ... ^ kCheckShares[7]` so that xor'ing all shares with
+ * the initial value of `error`, i.e. `kErrorBootDataInvalid`, and
+ * `kBootDataIdentifier` produces `kErrorOk`.
+ *
+ * Encoding generated with
+ * $ ./util/design/sparse-fsm-encode.py -d 6 -m 8 -n 32 \
+ *     -s 49105412 --language=c
+ *
+ * Minimum Hamming distance: 12
+ * Maximum Hamming distance: 23
+ * Minimum Hamming weight: 13
+ * Maximum Hamming weight: 20
+ */
+static const uint32_t kCheckShares[kHmacDigestNumWords + 1] = {
+    0xe021e1a9, 0xf81e8365, 0xbf8322db, 0xc7a37080, 0x271a933f,
+    0xdd8ce33f, 0x7585d574, 0x951777af, 0x381dee3a,
+};
+
+/**
+ * Checks whether the digest of a boot data entry is valid.
+ *
+ * @param boot_data A buffer that holds a boot data entry.
+ * @return Whether the digest of the entry is valid.
+ */
+rom_error_t boot_data_check(const boot_data_t *boot_data) {
+  static_assert(offsetof(boot_data_t, digest) == 0,
+                "`digest` must be the first field of `boot_data_t`.");
+
+  rom_error_t error = kErrorBootDataInvalid;
+  hmac_digest_t act_digest;
+  boot_data_digest_compute(boot_data, &act_digest);
+
+  size_t i = 0;
+  for (; launder32(i) < kHmacDigestNumWords; ++i) {
+    error ^=
+        boot_data->digest.digest[i] ^ act_digest.digest[i] ^ kCheckShares[i];
+  }
+  HARDENED_CHECK_EQ(i, kHmacDigestNumWords);
+  error ^= boot_data->identifier ^ kCheckShares[kHmacDigestNumWords];
+
+  return error;
```

Comment:
I forgot here that `error` could end up in a state which is not easy to map to a specific module.  Are we ok with that? Otherwise we would have to do something like

```cc
if (error != kErrorOk) {
  return kErrorBootDataInvalid;
}
return error;
```

Reply:
I was on the fence about this one: if the check fails in `boot_data_read()` we return `kErrorBootDataNotFound` and if the read succeeds, it shouldn't fail in mask_rom_boot(). But if it fails for any reason, it would be hard to debug. We already have this in sigverify so I also added it here. Thanks!